src/wp-includes/media.php

```diff
@@ -1866,6 +1869,79 @@ function wp_filter_content_tags( $content, $context = null ) {
 	return $content;
 }
 
+/**
+ * Use alternate mime type images in the content output when available.
+ *
+ * @since 6.0.0
+ *
+ * @param string $image         The HTML `img` tag where the attribute should be added.
+ * @param int    $attachment_id The attachment ID.
+ * @return string Converted `img` tag with `loading` attribute added.
+ */
+function wp_image_use_alternate_mime_types( $image, $attachment_id ) {
+	$metadata = wp_get_attachment_metadata( $attachment_id );
+	if ( empty( $metadata['file'] ) ) {
+		return $image;
+	}
+
+	$target_mimes = array( 'image/webp', 'image/jpeg' );
+
+	/**
+	 * Filter the content image mime type output selection and order.
+	 *
+	 * When outputting images in the content, the first mime type available will be used.
+	 *
+	 * @since 6.0.0
+	 *
+	 * @param array $target_mimes  The image output mime type and order. Default is array( 'image/webp', 'image/jpeg' ).
+	 * @param int   $attachment_id The attachment ID.
+	 * @return array The filtered output mime type and order. Return an empty array to skip mime type substitution.
+	 */
+	$target_mimes = apply_filters( 'wp_content_image_mimes', $target_mimes, $attachment_id );
+
+	if ( false === $target_mimes ) {
+		return $image;
+	}
+	// Find the appropriate size for the provided URL in the first available mime type.
+	foreach ( $target_mimes as $target_mime ) {
+		if ( ! isset( $metadata['sources'][ $target_mime ] ) || empty( $metadata['sources'][ $target_mime ]['file'] ) ) {
+			continue;
+		}
+
+		// Handle sub-sized image replacement.
+		foreach ( $metadata['sizes'] as $name => $size_data ) {
+			// Not the size we are looking for.
+			if ( empty( $size_data['file'] ) ) {
+				continue;
+			}
+			if ( empty( $size_data['sources'][ $target_mime ]['file'] ) ) {
+				continue;
+			}
+			$src_filename = wp_basename( $size_data['file'] );
+
+			// This is the same as the file we want to replace nothing to do here.
+			if ( $size_data['sources'][ $target_mime ]['file'] === $src_filename ) {
+				continue;
+			}
+
+			// Found a match, replace with the new filename and stop searching.
+			$image = str_replace( $src_filename, $size_data['sources'][ $target_mime ]['file'], $image );
+			continue;
+		}
+
+		// Handle full size image replacement.
+		$src_filename = wp_basename( $metadata['file'] );
+
+		// This is the same as the file we want to replace nothing else to do here.
+		if ( $metadata['sources'][ $target_mime ]['file'] === $src_filename ) {
+			return $image;
+		}
+
+		$image = str_replace( $src_filename, $metadata['sources'][ $target_mime ]['file'], $image );
+	}
```

Comment:
I think the implementation here is a bit off, mostly because it assumes the primary image format in the metadata is the one to replace, which is not necessarily the case (basically the `$src_filename` here is incorrectly determined).

The plugin implementation covers this more reliably. We shouldn't rely on what is the primary version of the image for the source format. For now, we should only ever replace JPEG images, potentially we can include a filter to expand that. But the source formats to possibly replace have to be separately specified somewhere (the plugin right now has a hard-coded JPEG in there).

Also, the current implementation here is running all of this logic even for images where it is irrelevant - for example in case of a PNG image. For example, a good idea would be to check somewhere at the start of this function whether the image is in any of the valid "source formats" - because if it isn't, we already know we don't need to make any replacement.

Reply:
I added some additional checks here for which source image mime types are replaceable which I currently hard coded to jpeg, we can make that more dynamic as you are proposing on the plugin .

The approach here is to go thru the data in the sizes array, and replace the file with the new mime type file. The urls stored in file should exactly match the markup urls, the code relies on the meta data instead of the markup though so it is a a little different that the plugin. 

I am assuming the "primary image format in the metadata is the one to replace" - when would that not be the case?